src/ralph/virtual/management/commands/openstack_sync.py

```diff
@@ -0,0 +1,532 @@
+                                               self.DATETIME_FORMAT),
+                )
+                self._save_object(new_server, 'add server %s'
+                                  % new_server.hostname)
+                new_server.tags.add(servers[server_id]['tag'])
+                self.summary['new_instances'] += 1
+
+                for ip in server['ips']:
+                    self._add_ip(new_server, ip, False)
+
+    # remove servers
+    def _cleanup_servers(self, servers, project_id):
+        """Remove servers which no longer exists in openstack project"""
+        try:
+            for server_id in (
+                self.ralph_projects[project_id]['servers'].keys(
+                ) - servers.keys()
+            ):
+                self._delete_object(CloudHost.objects.get(host_id=server_id))
+                self.summary['del_instances'] += 1
+        except KeyError:
+            pass
+
+    def _add_project(self, data, project_id):
+        """Add/modify project in ralph"""
+        if project_id in self.ralph_projects.keys():
+            project = CloudProject.objects.get(project_id=project_id)
+            if (
+                self.ralph_projects[project_id]['name'] != data['name'] or
+                self.openstack_projects[project_id]['tag']
+                    not in self.ralph_projects[project_id]['tags']
+            ):
+                self.summary['mod_projects'] += 1
+                project.cloudprovider = self.cloud_provider
+                project.name = data['name']
+                self._save_object(project, 'Modify name')
+                project.tags.add(data['tag'])
+
+        else:
+            self.summary['new_projects'] += 1
+            project = CloudProject(
+                name=data['name'],
+                project_id=project_id,
+                cloudprovider=self.cloud_provider,
+            )
+            self._save_object(project, 'Add project %s' % project.name)
+            project.tags.add(data['tag'])
+        self._add_servers(data['servers'], project)
+        self._cleanup_servers(data['servers'], project.project_id)
+
+    def _add_components(self, flavor, flavor_obj):
+        """add/remove components"""
+        names = [self._add_cpu(flavor, flavor_obj),
+                 self._add_ram(flavor, flavor_obj),
+                 self._add_disk(flavor, flavor_obj)]
+        try:
+            flavor_id = flavor_obj.flavor_id
+            ralph_names = self.ralph_flavors[flavor_id]['components']
+        except KeyError:
+            ralph_names = set()
+
+        for component in set(ralph_names) - set(names):
+            model = ComponentModel.objects.get(name=component)
+            self._delete_object(GenericComponent.objects.get(
+                asset=flavor_obj,
+                model=model))
+
+        if set(ralph_names) != set(names):
+            return 1
+        else:
+            return 0
+
+    def _add_cpu(self, flavor, flavor_obj):
```

Comment:
`_add_cpu`, `_add_ram` and `_add_disk` methods looks very similar - could you create one base method and call it from `_add_cpu`, `_add_ram` and `_add_disk` only passing particular params values (ex. `_add_component(flavor['cores'], 'cores vCPU', ComponentType.processor)`?